### PR TITLE
Added SaveSynchronizer and Related Tests

### PR DESCRIPTION
**Pull request by bmcelhanon**
Comprehensive Exercise completed! 

If you pull this down and run the tests, no internet connection is required. However, if you wish to actually test its functionality, please be sure to be on the student wifi network before summer break so as to be able to contact my FTP server.

Alternatively, you can set up your own server with FileZilla, modify the configuration XML to enable hashing, put a "Final" folder in the root directory with at least one empty file, and then recompile after changing the connection details in program.cs. 

Testing coverage isn't as wide as would be liked, but FTPServerHelper can't easily be tested without relying on the on-the-fly construction of an FTP Server with hashing support, so I'm assuming that when this is deployed the backend works. Additionally, ApplicationExecutor doesn't take well to testing either. If this doesn't work however, there is only a single five-line method in which something could have gone wrong. Should be extraordinarily easy to track down. Everything that could be offline tested has been tested.

Hashing was implemented to check if the remote file was updating while your local file was in use.
RegEx was worked in via text-based confirmation of wanting to overwrite a server file with a local file.
Serialization is used by the Settings class to manage the settings of the LocalSaveHandler.

Everything else should be easy enough to find. 

Thanks for everything!

### CODE REVIEW

File: SaveSynchronizer/SaveSynchronizer/LocalSaveHandler.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SaveSynchronizer
+{
+    class LocalSaveHandler
```
Comment (mattandccu) [inline SaveSynchronizer/SaveSynchronizer/LocalSaveHandler.cs:10]: Missed the XML comments on this class (-1)

**Review by mattandccu — COMMENTED** [review]
Great job creating a solution for a real-world scenario by combining technologies and techniques we discussed in class with other technologies you found in your own research. The organization and understandability of your code is top notch! 

I didn't see any examples of Collection usage, but I may have missed that. If I did, please just point me to that and I'll adjust accordingly.

Thanks for applying yourself to this final project and to the class throughout the semester.

58/60 = 97%

**Comment by bmcelhanon** [comment]
Thanks for the review!

I'm not sure it counts, but collections usage can be found very briefly on lines 68 & 69 of FtpSaveHandler.cs

I use the FTP library to pull down all the FTPListItems in the final directory, briefly creating a collection of FTPListItem. I then use a Lambda expression to select the most recent by modified descending, and the collection is not used again.

**Comment by mattandccu** [comment]
Thanks for the update. Lines 68 and 69 actually use an array of FtpListItem
rather than a collection.

Great job this semester! Thank you for all of your work and for continually
seeking to understand and learn new things.

Passing the MS exam was as huge accomplishment. Very well done!

I hope you have a great summer!

Matt

**Pull request closed without merging**